core/src/test/scala/org/apache/spark/shuffle/sort/io/LocalDiskShuffleMapOutputWriterSuite.scala

```diff
@@ -0,0 +1,154 @@
+import org.apache.spark.shuffle.IndexShuffleBlockResolver
+import org.apache.spark.util.Utils
+
+class LocalDiskShuffleMapOutputWriterSuite extends SparkFunSuite with BeforeAndAfterEach {
+
+  @Mock(answer = RETURNS_SMART_NULLS)
+  private var blockResolver: IndexShuffleBlockResolver = _
+
+  @Mock(answer = RETURNS_SMART_NULLS)
+  private var shuffleWriteMetrics: ShuffleWriteMetrics = _
+
+  private val NUM_PARTITIONS = 4
+  private val data: Array[Array[Byte]] = (0 until NUM_PARTITIONS).map { p =>
+    if (p == 3) {
+      Array.emptyByteArray
+    } else {
+      (0 to p * 10).map(_ + p).map(_.toByte).toArray
+    }
+  }.toArray
+
+  private val partitionLengths = data.map(_.length)
+
+  private var tempFile: File = _
+  private var mergedOutputFile: File = _
+  private var tempDir: File = _
+  private var partitionSizesInMergedFile: Array[Long] = _
+  private var conf: SparkConf = _
+  private var mapOutputWriter: LocalDiskShuffleMapOutputWriter = _
+
+  override def afterEach(): Unit = {
+    try {
+      Utils.deleteRecursively(tempDir)
+    } finally {
+      super.afterEach()
+    }
+  }
+
+  override def beforeEach(): Unit = {
+    MockitoAnnotations.initMocks(this)
+    tempDir = Utils.createTempDir()
+    mergedOutputFile = File.createTempFile("mergedoutput", "", tempDir)
+    tempFile = File.createTempFile("tempfile", "", tempDir)
+    partitionSizesInMergedFile = null
+    conf = new SparkConf()
+      .set("spark.app.id", "example.spark.app")
+      .set("spark.shuffle.unsafe.file.output.buffer", "16k")
+    when(blockResolver.getDataFile(anyInt, anyInt)).thenReturn(mergedOutputFile)
+    when(blockResolver.writeIndexFileAndCommit(
+      anyInt, anyInt, any(classOf[Array[Long]]), any(classOf[File])))
+      .thenAnswer { invocationOnMock =>
+        partitionSizesInMergedFile = invocationOnMock.getArguments()(2).asInstanceOf[Array[Long]]
+        val tmp: File = invocationOnMock.getArguments()(3).asInstanceOf[File]
+        if (tmp != null) {
+          mergedOutputFile.delete()
+          tmp.renameTo(mergedOutputFile)
+        }
+        null
+      }
+    mapOutputWriter = new LocalDiskShuffleMapOutputWriter(
+      0,
+      0,
+      NUM_PARTITIONS,
+      shuffleWriteMetrics,
+      blockResolver,
+      conf)
+  }
+
+  test("writing to an outputstream") {
+    (0 until NUM_PARTITIONS).foreach { p =>
+      val writer = mapOutputWriter.getPartitionWriter(p)
+      val stream = writer.openStream()
+      data(p).foreach { i => stream.write(i) }
+      stream.close()
+      intercept[IllegalStateException] {
+        stream.write(p)
+      }
+      assert(writer.getNumBytesWritten === data(p).length)
+    }
+    verifyWrittenRecords()
+  }
+
+  test("writing to a channel") {
+    (0 until NUM_PARTITIONS).foreach { p =>
+      val writer = mapOutputWriter.getPartitionWriter(p)
+      val outputTempFile = File.createTempFile("channelTemp", "", tempDir)
+      val outputTempFileStream = new FileOutputStream(outputTempFile)
```

Comment:
`Files.write(Path, byte[])` is basically a one-line version of these statements.